qt6: init at 6.2.0

milahu opened the pull request:
closed in favor of #169296

#### Motivation for this change

add latest version of qt, compile from source

#### Status

all builds are passing
libraries are usable in cmake projects
qmake support is not implemented, but has low priority (qmake is deprecated, qmake2cmake)

estimated build times on 4x2.4GHz

```
 80 min qtdeclarative
 40 min qtbase
 20 min qtwayland
720 min qtwebengine = 12 hours, similar to chromium

 [todo] pyside - takes surprisingly long
        C++ is heavy to parse, maybe caching could be better -> upstream issue
```

#### build

```
cd path/to/nixpkgs
time nix-build . -A qt6.qtbase --keep-failed 2>&1 | tee qt6.qtbase.$(date +%s).log
ls -d /nix/store/*-qtbase-6.2.?{,-{dev,bin}}
```

#### qt6 apps

https://github.com/milahu/awesome-qt6

#### contributing

feel free to help me with this beast ...

common problems (and solutions) with packaging qt6 for nixos: [packaging-qt6.md](https://github.com/milahu/nixpkgs/blob/qt-6-init/pkgs/development/libraries/qt-6/packaging-qt6.md)

#### related

* based on [the code for qt5](https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/libraries/qt-5/5.15)
* closes #108008
* gentoo https://github.com/gentoo/qt/pull/224
* arch linux https://archlinux.org/groups/x86_64/qt6/
* alpine linux https://gitlab.alpinelinux.org/alpine/aports/-/find_file/master

#### todo

* [ ] fix cycle error in qtvirtualkeyboard between $bin and $out (low priority)
  * `qtvirtualkeyboard` is currently disabled in `qt-6/6.2/default.nix`
* [ ] dynamic qt.conf for qmake. this was partly implemented for qt5 in [pkgs/development/libraries/qt-5/qt-env.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/qt-5/qt-env.nix#L13). ideally, qt.conf should be derived from buildInputs
  * low priority, because  qmake is deprecated in favor of cmake. we also have a conerter tool, `cmake2qmake` (in old versions this was called `pro2cmake.py`)
  * [ ] where is `cmake2qmake` currently? qtbase? qtbase.dev?
  * [ ] is this documented in the qt-6/*.md files?
* [x] libraries are usable in cmake projects. tested by compiling `qimgv` with qt6. qimgv needs qtbase, qtdeclarative, qtsvg
* [ ] move non-qt6 changes to separate PRs? (low priority)
   * [ ] qt6ct
   * [ ] qarma
   * [ ] qimgv
   * [ ] strawberry
   * [x] fetch-kde-qt.sh #148121 
   * [ ] openvg: shivavg, monkvg, amanithvg, ... (low priority)
   * [ ] openvg-headers (low priority)
   * [x] fix ninja line-clearing: `export TERM=dumb`
     * already fixed in #149810 
* [ ] switch apps to qt6
  * [x] qarma (gui dialogs generator)
  * [x] qimgv (image viewer)
  * [x] strawberry (music player)
  * [x] tdesktop (telegram desktop), done in https://github.com/NixOS/nixpkgs/pull/141883#issuecomment-1101495882
  * more apps ...?
* [ ] fix pending reviews (thanks sandro)
* [ ] squash all commits for qt6
* [ ] fix sqlite in qtbase? -> strawberry config: `CMake Warning at CMakeLists.txt:536 (message): The Qt sqlite driver test failed.`
* [x] pyside6 = qt6 bindings for python
  * [ ] fix `cmake/PySideHelpers.cmake` to detect qt modules in `QT_ADDITIONAL_PACKAGES_PREFIX_PATH`, which is set in `qtModule.nix` setupHook
* [ ] pyqt6 = qt6 bindings for python
  * last i checked, pyqt5 was bundled with qt *.so files. ideally link to qt packages
  * pyqt is a 3rd party library -> low priority
* [ ] qtbase tests and examples. how to build these? configureFlags or cmakeFlags? where are the tests? how to run the tests?
* [x] build all modules from srcs.nix (maybe some are missing)
  * all except qtvirtualkeyboard -> cycle error 
* [x] in wrapQtAppsHook, set `QML2_IMPORT_PATH` for qml apps (qtdeclarative)
  * already done in `pkgs/development/libraries/qt-6/hooks/wrap-qt-apps-hook.sh`
* [ ] update readme.md for qt6
* [ ] remove review code like qimgv-qt6
* [ ] patch-cmake-files.sh
   * [ ] should run before configurePhase &rarr; throw early on error
   * [x] add unit test
* [ ] update to latest qt version. currently 6.3.0 see https://download.qt.io/official_releases/qt/
  * [x] done in https://github.com/NixOS/nixpkgs/pull/141883#issuecomment-1101495882
    * [ ] qtwebengine 6.3.0
* [ ] update readme
  * [ ] maybe move readme to /docs/language-frameworks
* [ ] add version bools `qtbase.isQt6` and `qtbase.isQt5` etc
* [ ] fix `QT_PLUGIN_PATH`, see https://github.com/NixOS/nixpkgs/pull/141883#issuecomment-1101521629
 
#### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
qt-6: cp -r qt-5 qt-6; rm -rf 5.12 5.14; mv 5.15 6.2

Commit:
qt-6: find . -type f -exec sed -i 's|5.15.2|6.2.0|g; s|5.15|6.2|g; s|qt-5|qt-6|g' '{}' \;

Commit:
fetch-kde-qt.sh: use mirrorlist files to get sha256 sums

Commit:
qt-6: run fetch-kde-qt.sh to update srcs.nix

Commit:
qt-6: update all-packages.nix, add empty qt6-packages.nix

Commit:
qt-6: qtbase: fix patches

milahu commented:
@SuperSandro2000 thanks for spamming the pr ; ) this would be easier as pr to [my branch](https://github.com/milahu/nixpkgs/tree/qt-6-init), but nevermind .. next time

Commit:
qt-6: lint nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
qt-6: fix out-of-tree build LD_LIBRARY_PATH

Commit:
qt-6: fix syntax in modules/qtbase.nix

Commit:
qt-6: fix: Qt5Core.pc: No such file

Commit:
qt-6: remove missing modules

Commit:
qt-6: preBuild: add note: build is silent

Commit:
qt-6: lint qt6-packages.nix

milahu commented:
maybe rename?

```diff
-  pkgs/development/libraries/qt-6
+  pkgs/development/libraries/qt6
```

Commit:
openvg-headers: init 1.1 (TODO remove)

Commit:
openvg.shivavg: init 0.2.1-e0da003 [TODO remove]

Commit:
qt-6.qtbase: enable most features WIP DRAFT

TODO enable openvg, allow to choose openvg.shivavg or openvg.monkvg or ...
currently, cmake does not find libOpenVG.so

Commit:
openvg.monkvg: init 2a04624

Commit:
qt-6.qtbase: force opencv WIP DRAFT

opencv maybe not implemented in qt6

also: enable more features in qtbase

Commit:
qt6.qtbase fix build? WIP DRAFT

Commit:
qt6.qtbase: cache buildPhaseResult to help debugging

milahu commented on pkgs/development/libraries/qt-6/modules/qtbase.nix:
```diff
@@ -181,6 +181,10 @@ stdenv.mkDerivation {
   '';
 
+  preBuild = ''
+    echo 'note: the build is silent. please wait'
```
`ninja -d explain` could be another "reasonable default" for ninja

this makes ninja explain the "why?" of its actions
for example

```
ninja -d explain install
ninja explain: output build.ninja older than most recent input CMakeFiles/3.21.2/CMakeCCompiler.cmake (1635672439927832152 vs 1635672447966884550)
[0/1] Re-running CMake...
```

ttuegel commented on pkgs/development/libraries/qt-6/modules/qtbase.nix:
```diff
@@ -0,0 +1,451 @@
+    echo "postInstall: find mkspecs"
+    find mkspecs
+    echo "postInstall: find build/mkspecs"
+    find build/mkspecs
```
With respect to the linked issue, the correct way to find the interpreter in a Nix build is `cat $NIX_CC/nix-support/dynamic-linker`. Is it worth it to upstream such a trivial patch, versus using `substituteInPlace`, when the solution doesn't generalize to other platforms anyway?

milahu replied:
> the correct way to find the interpreter in a Nix build is `cat $NIX_CC/nix-support/dynamic-linker`

you mean this patch?

```diff
-  readelf -l /bin/ls
+  readelf -l $(which ls)

-  /bin/pwd
+  pwd
```

how is `ld-linux.so` relevant here?

> Is it worth it to upstream such a trivial patch

yepp, cos its an upstream bug : )
calls like `/bin/pwd`, `/bin/ls`, `/bin/rm` are not portable, and not needed in shell scripts
but no worry, thats not the reason why my draft is moving slow ...

currently, im splitting buildPhase and installPhase into 2 derivations
since buildPhase takes 2 hours (on my machine) and installPhase takes like 2 minutes ...
so debugging the installPhase is a terrible feedback loop
i hope i can generalize this build/install splitting for other large builds.
once the installPhase is working, the splitting can be disabled

with the `ninja` build tool, this is a challenge,
cos ninja validates files and paths before install (on failure calls `rerun_cmake`),
and i have to patch file times (and maybe checksums) ...
this should be easier with `cmake` (but its slower than ninja, so they say)

milahu replied:
> currently, im splitting buildPhase and installPhase into 2 derivations

done in 3e50ae7 : )

Commit:
qt6.qtbase: implement splitBuildInstall, make qtbase build

splitBuildInstall:
split buildPhase and installPhase into separate derivations
so we can debug installPhase without re-running buildPhase

Commit:
qt6: fix _IMPORT_PREFIX paths in cmake files

Commit:
qt6.qtModule: qmake -> cmake

Commit:
qt6.qtwayland: init

Commit:
qt6: add dependency cmake

Commit:
qarma: init 2021-10-05

Commit:
qt6.qtsvg: init

Commit:
qt6.qt5compat: init

Commit:
qt6.qtModule: replace _IMPORT_PREFIX with nix output paths

Commit:
qimgv: qt5to6 WIP runtime error: plugin wayland is not found

milahu commented on pkgs/development/libraries/qt-6/qtModule.nix:
```diff
@@ -48,6 +51,43 @@ mkDerivation (args // {
         done
     fi
 
+    # TODO refactor. same code in qtbase.nix
```
TODO refactor. same code in qtbase.nix

milahu commented on pkgs/development/libraries/qt-6/modules/qtbase.nix:
```diff
@@ -0,0 +1,534 @@
+    # TODO how is this working in qt5?
+    set -o xtrace
+    mkdir $bin
+    cp -r --verbose bin $bin
```
qtbase build is passing

milahu commented on pkgs/development/libraries/qt-6/modules/qtbase.nix:
```diff
@@ -59,13 +61,16 @@ let
 
-  cacheBuildPhaseResult = true; # save the result of buildPhase in a separate derivation
-  # make easier to debug postInstall, postFixup
+  splitBuildInstall = true;
```
TODO generalize this for `stdenv.mkDerivation`

useful for debugging large builds = buildPhase runs longer than 5 minutes

the option `splitBuildInstall` can be enabled to get a "development mode" for a package,
and as soon as the package is "ready for production",
we remove `splitBuildInstall` to run buildPhase and installPhase in one derivation

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/splitbuildinstall-split-buildphase-and-installphase-for-large-packages/15874/1

Commit:
qt6.qtModule: move plugins to $bin output

Commit:
qt6.qtwayland: fix for new qtModule interface TODO rebase: fixup to e58d329

Commit:
qt6.qtsvg: fix for new qtModule interface TODO rebase: fixup to 15d1548

Commit:
qt6.qtimageformats: init

Commit:
qimgv: add dependency wayland TODO rebase: fixup to 5f7fb21

Commit:
qt6.qtwayland: add dependencies TODO rebase: fixup to e58d329

Commit:
qt6.qtbase, qt6.qtModule: fix install paths

milahu commented:
first app is working : )

```
nix-build . -A qimgv
./result/bin/qimgv
```

SuperSandro2000 commented on pkgs/tools/misc/qarma/default.nix:
```diff
@@ -0,0 +1,36 @@
+    sha256 = "KFoFywFeGqNmE1y49DrXJZ1jIK5jMOCOspkkFME+DR8=";
+  };
+
+  patchPhase = ''
```
Please switch this to postPatch.

SuperSandro2000 commented on pkgs/tools/misc/qarma/default.nix:
```diff
@@ -0,0 +1,36 @@
+  version = "2021-10-05";
+
+  src = fetchFromGitHub {
+    name = "${pname}-${version}-source";
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/pyside6/default.nix:
```diff
@@ -0,0 +1,70 @@
+  configurePhase = ":";
+
+  buildPhase = ''
+    echo buildPhase
```
```suggestion
```

milahu replied:
WIP

SuperSandro2000 commented on pkgs/development/python-modules/pyside6/default.nix:
```diff
@@ -0,0 +1,70 @@
+  '';
+
+  installPhase = ''
+    echo installPhase
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/pyside6/default.nix:
```diff
@@ -0,0 +1,70 @@
+
+  buildPhase = ''
+    echo buildPhase
+    python setup.py build
```
```suggestion
    ${python.interpreter} setup.py build
```

SuperSandro2000 commented on pkgs/development/python-modules/pyside6/default.nix:
```diff
@@ -0,0 +1,70 @@
+
+  installPhase = ''
+    echo installPhase
+    python setup.py install
```
```suggestion
    ${python.interpreter} setup.py install
```

SuperSandro2000 commented on pkgs/development/python-modules/pyside6-tools/default.nix:
```diff
@@ -0,0 +1,52 @@
+  ];
+
+  postPatch = ''
+    cd sources/${pname}
```
This should probably be replaced with sourceRoot.

milahu replied:
yepp, also in qtbase ... WIP

SuperSandro2000 commented on pkgs/development/libraries/qt-6/qtModule.nix:
```diff
@@ -0,0 +1,154 @@
+    ${args.postFixup or ""}
+  '';
+
+  meta = {
```
```suggestion
  meta = with lib; {
```

SuperSandro2000 commented on pkgs/development/libraries/qt-6/qtModule.nix:
```diff
@@ -0,0 +1,154 @@
+{ lib, mkDerivation, perl, cmake }:
+
+let inherit (lib) licenses maintainers platforms; in
+
+{ self, srcs, patches }:
```
```suggestion
{ self, srcs, patches }:
```

SuperSandro2000 commented on pkgs/development/libraries/qt-6/modules/qtwebkit.nix:
```diff
@@ -0,0 +1,73 @@
+qtModule {
+  pname = "qtwebkit";
+  qtInputs = [ qtbase qtdeclarative qtlocation qtsensors ]
+    ++ lib.optional (stdenv.isDarwin && lib.versionAtLeast qtbase.version "5.9.0") qtmultimedia
```
```suggestion
    ++ lib.optional (stdenv.isDarwin) qtmultimedia
```

milahu replied:
qt6.qtwebkit does not-yet exist, sorry for the confusion

SuperSandro2000 commented on pkgs/development/libraries/qt-6/modules/qtwebkit.nix:
```diff
@@ -0,0 +1,73 @@
+    '';
+    buildInputs = [ perl ];
+  };
+  usingAnnulenWebkitFork = lib.versionAtLeast qtbase.version "5.11.0";
```
```suggestion
```
This can be removed, right?

SuperSandro2000 commented on pkgs/development/libraries/qt-6/modules/qtwebkit.nix:
```diff
@@ -0,0 +1,73 @@
+, qtbase, qtdeclarative, qtlocation, qtmultimedia, qtsensors, qtwebchannel
+, fontconfig, libwebp, libxml2, libxslt
+, sqlite, systemd, glib, gst_all_1, cmake
+, bison, flex, gdb, gperf, perl, pkg-config, python2, ruby
```
can we use python3?

SuperSandro2000 reviewed (commented):
Can we re-use the hooks from 5 if they did not change?

The README needs to be updated and probably should be moved to /docs/language-frameworks.

Commit:
qt6.qtModule: remove hasPlugin interface TODO rebase fixup a67bd4f

some packages have no $out -> simply set outputs

Commit:
qt6.qtModule: refactor meta TODO rebase fixup a67bd4f

Commit:
fetch-kde-qt.sh: print output file on success TODO rebase fixup 753f8f64c

Commit:
qt6.qtModule: fatal error is only warning TODO rebase fixup a67bd4f

Commit:
qt6.qtshadertools: init 6.2.0

Commit:
qt6.qtdeclarative: init 6.2.0

Commit:
qt6.qtimageformats: remove hasPlugins TODO rebase fixup e39bd1ec8

Commit:
qt6.qtsvg: remove hasPlugins TODO rebase fixup e8fb605f3

Commit:
qt6.qtwayland: remove hasPlugins TODO rebase fixup 67fb0e6a2

Commit:
qt6.qtbase: use $sourceRoot TODO rebase fixup 9cc40cd 3e50ae7 fc5a296 c17158f bc22b6d 471c566 173eee7 9bafca5

Commit:
qt6.qtshadertools: add to modules TODO fixup 8c03121

Commit:
qt6: remove all patches from qt5

Commit:
qt6.qtquick3d: init 6.2.0

Commit:
qt6.qtquicktimeline: init 6.2.0

Commit:
qt6.qt5compat: add qtdeclarative TODO fixup 2d110ab57

Commit:
qt6.qtModule: make patches optional TODO fixup ea39f7a

Commit:
qt6.qtbase: make patches optional

milahu commented on pkgs/development/libraries/qt-6/modules/qtdeclarative.nix:
```diff
@@ -1,9 +1,15 @@
+}:
 
 qtModule {
   pname = "qtdeclarative";
```
note: build is failing
build time is 2.5 hours on dual core
WIP splitBuildInstall

Commit:
openvg.shivavg: update src, lint nix TODO rebase fixup c2c3d25

Commit:
openvg-headers: lint nix TODO rebase fixup c019b5a

Commit:
qt6.qtModule: add splitBuildInstall

Commit:
qt6.qtdeclarative: enable splitBuildInstall TODO rebase remove

Commit:
maintainers: add milahu [add matrix] [TODO fixup?]

Commit:
qt6: 6.2.0 -> 6.2.1

Commit:
qt6.qtbase: fix deps, rm openvg

Commit:
qt6.qtModule: add maintainer milahu

Commit:
qt6.qtdeclarative: add splitBuildInstall, fix build

Commit:
qt6.qtwayland: fix deps

Commit:
qt6: bugfix in split-build-install.import.sh

milahu commented on pkgs/development/libraries/qt-6/modules/qtdeclarative.nix:
```diff
@@ -29,11 +39,160 @@ qtModule {
+    s+="s/\\\''${_IMPORT_PREFIX}\/(\.\/)?plugins/\\\''${_''${moduleNAME}_NIX_BIN}\/lib\/qt-${version}\/plugins/g;"
+    s+="s/\\\''${_IMPORT_PREFIX}\/(\.\/)?bin/\\\''${_''${moduleNAME}_NIX_DEV}\/bin/g;" # qmake ...
+    s+="s/\\\''${_IMPORT_PREFIX}\/(\.\/)?mkspecs/\\\''${_''${moduleNAME}_NIX_DEV}\/mkspecs/g;"
+    s+="s/\\\''${_IMPORT_PREFIX}\/(\.\/)?qml/\\\''${_''${moduleNAME}_NIX_OUT}\/lib\/qt-${version}\/lib\/qml/g;"
```
by patching the `qml` paths, the `qtdeclarative` build is passing
but `libQtQuick.so` is not usable from cmake projects

for example [qtquicktimeline](/home/user/src/nixos/nixpkgs-2021-10-12/nixpkgs/pkgs/development/libraries/qt-6/modules/qtquicktimeline.nix) requires QtQuick

```cmake
find_package(Qt6 6.2.1 QUIET CONFIG OPTIONAL_COMPONENTS Quick)

if(NOT TARGET Qt::Quick)
    message(NOTICE "Skipping the build as the condition \"TARGET Qt::Quick\" is not met.")
    return()
endif()

# -> Skipping the build as the condition "TARGET Qt::Quick" is not met.
```

per `cmakeFlags = [ "-DCMAKE_FIND_DEBUG_MODE=TRUE" ];`
this will use `qtbase-6.2.1-dev/lib/cmake/Qt6/Qt6Config.cmake`
but `Qt6Config.cmake` cannot find `qtdeclarative`

`CMAKE_LIBRARY_PATH` has `/nix/store/qhh0spm39dvi6xkg386721h3gsb4p9an-qtdeclarative-6.2.1/lib`

but we would need something like
`CMAKE_CONFIG_PATH` with `/nix/store/gslwy5b5pfvjdi8p43q2hklg6nqafjg4-qtdeclarative-6.2.1-dev/lib/cmake`

milahu replied:
next error:

```
  The imported target "Qt6::qmldom" references the file

     "/nix/store/gslwy5b5pfvjdi8p43q2hklg6nqafjg4-qtdeclarative-6.2.1-dev/bin/qmldom"

  but this file does not exist.  Possible reasons include:
```

... but the file is in `/nix/store/2s3nk9frssyib2qp29hp30i2i1500mp2-qtdeclarative-6.2.1/bin/qmldom`

&rarr; bug in `moveQtDevTools`?

&rarr; workaround in `qtdeclarative.nix`

```sh
mv $out/bin $dev/bin
```

Commit:
qt6.qtbase: fix configureFlags

Commit:
qt6.qtbase: patch $dev/mkspecs/features/qt_config.prf

TODO rebase: maybe this is fixed in ad2e36de8 fix configureFlags

Commit:
qt6.qtModule: update cmake patching

Commit:
qt6.qtbase: fix configureFlags

Commit:
qt6.qmake2cmake: init at 6.2.1, aka pro2cmake.py

Commit:
qarmaForQt6: init at 2021-10-05 [TODO rename]

Commit:
qt6.qmake2cmake: move nix file TODO rebase fixup to df416a5117

milahu commented on pkgs/tools/misc/qarma-for-qt6/default.qmake.nix:
```diff
@@ -0,0 +1,139 @@
+    find /tmp/.config
+    cat /tmp/.config/QtProject/QMake.conf
+
+    echo QT_INSTALL_HEADERS
+    echo "expected = ${qtbase.dev}/include"
+    echo "actual   = $(qmake -query QT_INSTALL_HEADERS)"
```
qmake is failing here

```
cd nixpkgs
nix-build . -A qarmaForQt6Qmake

expected = /nix/store/26k5lszsa2adc2qxj5yqxc06ygbq0zf3-qtbase-6.2.1-dev/include
actual   = /nix/store/52p7a812jmy3fl63f349dq1mwlai3xir-qtbase-6.2.1/include
```

qmake expects qt headers in qtbase $out, but headers are in qtbase $dev

`qmake2cmake` aka [pro2cmake.py](https://code.qt.io/cgit/qt/qtbase.git/tree/util/cmake/pro2cmake.py) is a workaround to build qmake projects
example project is [qarma](https://github.com/milahu/nixpkgs/blob/qt-6-init/pkgs/tools/misc/qarma-for-qt6/default.nix) &rarr; ```nix-build . -A qarmaForQt6```

milahu replied:
should be fixed in c5928ecb52c81bb8aeb5fe37591d74b8bf519668

at least `qmake -query QT_INSTALL_HEADERS` is now returning the correct value

Commit:
qarmaForQt6: update patch TODO fixup 5f48ab435

milahu commented on pkgs/tools/misc/qarma-for-qt6/default.nix:
```diff
@@ -1,6 +1,11 @@
 qt6 build is failing
 https://github.com/luebking/qarma/issues/41
+
+FIXME broken cmake files in qtbase?
+  #include <QX11Application>
+  QX11Application: No such file or directory
+  should be provided by /nix/store/*-qtbase-6.2.1-dev/include/QtGui/qguiapplication_platform.h
```
cmake is failing here : /

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -27253,6 +27253,7 @@ with pkgs;
   pybitmessage = callPackage ../applications/networking/instant-messengers/pybitmessage { };
 
   qarma = libsForQt5.callPackage ../tools/misc/qarma { };
+  qarmaForQt6 = libsForQt6.callPackage ../tools/misc/qarma-for-qt6 { };
```
```suggestion
  qarma_qt6 = libsForQt6.callPackage ../tools/misc/qarma/qt6.nix { };
```
I would expect that it would be called like this. Also we can put the file into the same directory as the normal qarma.

milahu replied:
this is temporary, to compare qt5 and qt6 builds. goal is to move to qt6 by default

SuperSandro2000 replied:
Thats a huge task. Thanks for trying to do it! If you hit a roadbloack I would be totally fine for having Qt5 and Qt6 alongside each other for a while.

SuperSandro2000 commented on pkgs/tools/misc/qarma-for-qt6/default.qmake.nix:
```diff
@@ -0,0 +1,139 @@
+    "QT_INSTALL_HEADERS=${qtbase.dev}/include"
+  ];
+
+  buildInputs = [ qtbase qtbase.dev qt5compat qt5compat.dev ];
```
Does it really need to have the .dev package explicitly?

SuperSandro2000 commented on pkgs/tools/misc/qarma-for-qt6/default.qmake.nix:
```diff
@@ -0,0 +1,139 @@
+    ./qarma-qt6.patch
+  ];
+
+  patchPhase = ''
```
Currently the above patch is not applied.

milahu replied:
fixed in 591178c668a28882b402538d4365b7342b751b7b

SuperSandro2000 commented on pkgs/tools/misc/qarma-for-qt6/default.nix:
```diff
@@ -0,0 +1,58 @@
+    ./qarma-qt6.patch
+  ];
+
+  patchPhase = ''
```
The above patch is not applied and overriding patchPhase is a big nogo.

```suggestion
  postPatch = ''
```

SuperSandro2000 commented on pkgs/development/libraries/qt-6/modules/qtbase.nix:
```diff
@@ -389,24 +398,22 @@ qtbaseDrv = stdenv.mkDerivation rec {
 
-    "-make libs"
-    "-make tools"
+    "-make" "libs"
+    "-make" "tools"
```
Does that make a difference? It shouldnt.

milahu replied:
> Does that make a difference?

im afraid no (did not fix the bug)

> It shouldnt.

it should ^^

how would i pass `-o "a b"` to configure? `configureFlags = [ ''-o "a b"'' ]`...?

SuperSandro2000 replied:
> how would i pass -o "a b" to configure? configureFlags = [ ''-o "a b"'' ]...?

I think that would need to be escaped but I am not fully sure:

```nix
configureFlags = [ "-o \"a b\"" ]
```

SuperSandro2000 commented on pkgs/development/libraries/qt-6/modules/qmake2cmake.nix:
```diff
@@ -0,0 +1,68 @@
+  # use original name as alias
+  postInstall = ''
+    rm $out/bin/pro2cmake.py
+    ln -s ${pname} $out/bin/pro2cmake.py
```
This should be cp or we reference the sources at runtime and cannot garbage collect them.

milahu replied:
> This should be cp

wdym?

SuperSandro2000 replied:
Instead of symlinking (ln) we should copy (cp) or move (mv) the file so that the final system has no reference to the src anymore.

milahu replied:
> so that the final system has no reference to the src anymore.

there is no reference to src


<pre>
$ ls -l result/bin/
total 188
-r--r--r-- 1 root root  4886  1. Jan 1970  cmakeconversionrate.py
-r--r--r-- 1 root root  6513  1. Jan 1970  condition_simplifier_cache.py
-r--r--r-- 1 root root  9278  1. Jan 1970  condition_simplifier.py
-r--r--r-- 1 root root 54946  1. Jan 1970  configurejson2cmake.py
-r--r--r-- 1 root root 37268  1. Jan 1970  helper.py
-r--r--r-- 1 root root  3991  1. Jan 1970  json_parser.py
lrwxrwxrwx 1 root root    11  1. Jan 1970  <span color="#55FFFF"><b>pro2cmake.py</b></span> -&gt; <span color="#55FF55"><b>qmake2cmake</b></span>
-r--r--r-- 1 root root  8156  1. Jan 1970  pro_conversion_rate.py
-r-xr-xr-x 1 root root   516  1. Jan 1970  <span color="#55FF55"><b>qmake2cmake</b></span>
-r--r--r-- 1 root root 14832  1. Jan 1970  qmake_parser.py
-r--r--r-- 1 root root  8286  1. Jan 1970  run_pro2cmake.py
-r--r--r-- 1 root root   233  1. Jan 1970  setup.py
-r--r--r-- 1 root root 16254  1. Jan 1970  special_case_helper.py
</pre>

`pro2cmake.py` is just an alias for `qmake2cmake`
since `pro2cmake.py` is the original name of the script

```
$ ./result/bin/pro2cmake.py 2>&1 | head -n 1
usage: qmake2cmake [-h] [--debug] [--debug-parser] [--debug-parse-result]

$ ./result/bin/qmake2cmake 2>&1 | head -n 1
usage: qmake2cmake [-h] [--debug] [--debug-parser] [--debug-parse-result]
```

SuperSandro2000 commented on pkgs/development/libraries/qt-6/modules/qmake2cmake.nix:
```diff
@@ -0,0 +1,68 @@
+  sourceRoot = "qtbase-everywhere-src-${version}/util/cmake";
+
+  propagatedBuildInputs =
+    (with python3Packages; [ pyparsing portalocker sympy pytest ])
```
pytest should never be propagated because it makes it very hard to override it. If it is only required for checkPhase please move it to checkInputs, otherwise it goes to buildInputs.

milahu replied:
done. `sympy pytest` are test deps

Commit:
qt6.qmake2cmake: fix checkInputs TODO rebase fixup to df416a5117

Commit:
qt6.qmake2cmake: clean nix TODO rebase fixup to df416a5117

Commit:
qt6.qtbase: qmake: read config from env vars

Commit:
qt6.qtbase: actually fix qmake. undo fe75dce135

Commit:
qt6: 6.2.1 -> 6.2.2

Commit:
qt6.qtbase: actually fix qmake TODO fixup a562e71a

milahu commented on pkgs/development/libraries/qt-6/hooks/qmake-hook.sh:
```diff
@@ -0,0 +1,46 @@
+
+    echo "QMAKEPATH=$QMAKEPATH"
+    echo qmake "${qmakeFlags[@]}"
+    qmake "${qmakeFlags[@]}"
```
todo: run `qmake -qtconf /path/to/qt.conf` with a dynamic `qt.conf` file, generated from `buildInputs`

milahu commented on pkgs/development/libraries/qt-6/6.2/default.nix:
```diff
@@ -0,0 +1,203 @@
+
+      wrapQtAppsHook = makeSetupHook {
+        deps = [ self.qtbase.dev makeWrapper ]
+          ++ lib.optional stdenv.isLinux self.qtwayland.dev;
```
this adds a `qtwayland` dependency to all `wrapQtAppsHook` apps ... lets call it a feature

milahu commented on pkgs/applications/graphics/qimgv/default.nix:
```diff
@@ -10,6 +10,9 @@
 , mpv
 , opencv4
 , qtbase
+, libglvnd, libxkbcommon, vulkan-headers # TODO should be inherited from qtbase
```
dependencies are not propagated from qtbase to qimgv

```nix
# qtbase
{
  propagatedBuildInputs = [
    vulkan-headers
    libxkbcommon
    libGL # effectively the same as libglvnd
    # ...
  ];
```

```nix
# qimgv
{
  buildInputs = [
    qtbase
    # ...
  ];
```

nrdxp commented:
attempting to build WebEngine from the latest commit of this PR results in the following error:
```
❯ nix build .#qt6.qtwebengine
error: attribute 'qtlocation' missing

       at /nix/store/86r45zyhl672gcxlkkgb41ah7dsp22br-source/pkgs/development/libraries/qt-6/qtModule.nix:15:29:

           14|   inherit (args) pname;
           15|   version = args.version or srcs.${pname}.version;
             |                             ^
           16|   src = args.src or srcs.${pname}.src;
(use '--show-trace' to show detailed location information)
```

Commit:
qt6.qtwebsockets: init at 6.2.2

Commit:
qt6.qtwebchannel: init at 6.2.2

Commit:
qt6: add packaging-qt6.md

Commit:
qt6.qtwebengine: git -> tar.xz [DRAFT]

Commit:
qt6.qtdeclarative: fix cmake path to /bin/qmldom

milahu commented:
good news, there is a stable release for qtwebengine 6.2.2

> error: attribute 'qtlocation' missing

qtlocation is legacy from qt5, either it was removed in qt6, or it is not yet ported to qt6

starting compile of qtwebengine ... the source tarball has 350 megabytes 0__o

Commit:
qt6.qtwebengine: disable cmake trace [DRAFT]

Commit:
qt6.qtserialport: init at 6.2.2

Commit:
qt6.qtpositioning: init at 6.2.2

Commit:
qt6.qtwebengine: fix build [DRAFT]

Artturin commented:
@jonringer 's beast server would make the dev loop a lot faster

I suggest you make a pr in https://github.com/jonringer/server-configuration and see if he accepts it

milahu commented:
thanks. im using `ccache` with  success. i even get cache hits in "direct mode" = no C preprocessor

<details>

```
./build-qt6-qtsvg.sh
run 1: real 5m40.369s
run 2: real 1m51.376s
3x faster

./build-qt6-qtserialport.sh
run 1: real 1m31.229s
run 2: real 0m29.038s
3x faster

./build-qt6-qtshadertools.nix
run 1: real 18m5.639s
run 2: real 0m55.255s
18x faster
```

</details>

also works with `sccache` which allows distributed builds,
but i guess `qtwebengine` is too small to justify the complexity

Commit:
qt6.qtbase: cleanup [DRAFT] [TODO fixup]

Commit:
qt6.qtbase: cleanup [DRAFT] [TODO fixup] [2]

Artturin commented on pkgs/development/libraries/openvg/default.nix:
```diff
@@ -0,0 +1,113 @@
+# TODO add amanithvg
```
can openvg stuff be moved to a separate PR as its not required for qt-6?

Commit:
qt6.qtbase: fix compile [TODO fixup]

Artturin commented on pkgs/development/libraries/unixODBCDrivers/default.nix:
```diff
@@ -29,13 +29,13 @@
   mariadb = stdenv.mkDerivation rec {
     pname = "mariadb-connector-odbc";
-    version = "3.1.4";
+    version = "3.1.13";
```
this update can dropped because its already updated on master

Artturin commented:
remember to rebase this branch sometimes
here's my script to do that to master
```bash
#!/usr/bin/env bash
# if not done before 
# git remote add upstream https://github.com/NixOS/nixpkgs.git

git fetch upstream
git checkout master
git rebase upstream/master
git push
```

to update this branch you would add the remote and fetch it, then `git rebase upstream/master`

milahu commented:
> jonringer 's beast server would make the dev loop a lot faster

thanks, that was really helpful. some of the modules are really huge ...
qtwebengine takes 1 hour on 32x3.6GHz

> remember to rebase this branch sometimes

sure. should i force-push this branch or create a new branch?
i assume when i force-push, then reviewers must read everything again? cc @SuperSandro2000 etc

Artturin commented:
> > remember to rebase this branch sometimes
> 
> sure. should i force-push this branch or create a new branch?

Drop the odbc update commit and them update the branch and force push

> i assume when i force-push, then reviewers must read everything again? cc @SuperSandro2000 etc

You don't have to worry about that

SuperSandro2000 commented:
> sure. should i force-push this branch or create a new branch?

Just force push.

> i assume when i force-push, then reviewers must read everything again? cc @SuperSandro2000 etc

Na, I marked the read files as read and even if they change then it will be like this. It needs to be done or otherweise the jump might be to big and things are immediately broken with other package updates.

Commit:
qt6: fix all modules - TODO fixup

Commit:
qimgv: 1.0.1 -> 1.0.2

Commit:
qt6: update packaging-qt6.md

Commit:
qt6: nixpkgs-fmt

Commit:
qt6: fix all-packages.nix

milahu commented:
eval check is failing on darwin, no idea whats wrong

```
$ nix-build pkgs/top-level/release.nix -A tarball.nixpkgs-basic-release-checks --arg supportedSystems '[ "aarch64-darwin" "aarch64-linux" "x86_64-linux" "x86_64-darwin"  ]' --show-trace
this derivation will be built:
  /nix/store/pcsf6vfpq5yrzm4pv9rmvf8ki24lz4xm-nixpkgs-release-checks.drv
building '/nix/store/pcsf6vfpq5yrzm4pv9rmvf8ki24lz4xm-nixpkgs-release-checks.drv'...
checking Nixpkgs on aarch64-darwin
error: builder for '/nix/store/pcsf6vfpq5yrzm4pv9rmvf8ki24lz4xm-nixpkgs-release-checks.drv' failed with exit code 1;
       last 1 log lines:
       > checking Nixpkgs on aarch64-darwin
       For full logs, run 'nix log /nix/store/pcsf6vfpq5yrzm4pv9rmvf8ki24lz4xm-nixpkgs-release-checks.drv'.

$ nix log /nix/store/pcsf6vfpq5yrzm4pv9rmvf8ki24lz4xm-nixpkgs-release-checks.drv
checking Nixpkgs on aarch64-darwin
```

Commit:
revert maintainer-list.nix

Commit:
qt6: build apps with qt6: qarma, qimgv, strawberry

milahu committed:
qtbase: add QT_FEATURE_openssl_linked=ON

milahu committed:
qtModule: prettier log output

milahu committed:
qt6: nitpicks in packaging-qt6.md

milahu commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28292,6 +28296,7 @@ with pkgs;
   qjackctl = libsForQt5.callPackage ../applications/audio/qjackctl { };
 
   qimgv = libsForQt5.callPackage ../applications/graphics/qimgv { };
+  qimgv-qt6 = libsForQt6.callPackage ../applications/graphics/qimgv { };
```
just for testing

milahu committed:
qt6ct: init at 2021-12-22

NickCao commented:
Also: don't forget to set QML2_IMPORT_PATH in wrapQtAppsHook, or otherwise qml applications won't work

jonringer commented on pkgs/development/libraries/qt-6/6.2/default.nix:
```diff
@@ -0,0 +1,213 @@
+      };
+
+      # experiment
+      qtbaseDistbuild = callPackage ../modules/qtbase.distbuild.nix {
```
This is probably meant for reviewing purposes, but:

```
pkgs/development/libraries/qt-6/modules/qtbase.distbuild.nix': No such file or directory
```

milahu replied:
thanks. that was an experiment with distributed compiling

SuperSandro2000 commented on pkgs/tools/misc/qt6ct/default.nix:
```diff
@@ -0,0 +1,41 @@
+
+stdenv.mkDerivation rec {
+  pname = "qt6ct";
+  version = "2021-12-22";
```
```suggestion
  version = "unstable-2021-12-22";
```

SuperSandro2000 commented on pkgs/tools/misc/qarma/default.nix:
```diff
@@ -0,0 +1,39 @@
+
+mkDerivation rec {
+  pname = "qarma";
+  version = "2022-01-16";
```
```suggestion
  version = "unstable-2022-01-16";
```

SuperSandro2000 commented on pkgs/tools/misc/qarma/default.nix:
```diff
@@ -0,0 +1,39 @@
+
+  buildInputs = [
+    qtbase
+  ] ++ lib.optionals (lib.versions.major qtbase.version == "5") [
```
If we need this more often we should passthru it from qtbase.

milahu replied:
no, only some pkgs need this

SuperSandro2000 commented on pkgs/tools/misc/qarma/default.nix:
```diff
@@ -0,0 +1,39 @@
+, wrapQtAppsHook
+, qtbase
+, qmake
+, qtx11extras ? null # qt5
+, qt5compat ? null # qt6
```
We do not want to use ``? null`` in inputs.

milahu replied:
this is for compatibility with 

libsForQt5.callPackage
libsForQt6.callPackage

better solution?

prusnak replied:
You can just drop the `? null` since you are using `lib.optionals` below and nix has lazy evaluation.

milahu replied:
> You can just drop the `? null`

nope:

> $ nix-build ./nixpkgs/ -A qarma-qt6
> error: anonymous function at /tmp/milahu/nixpkgs-qt6-init/nixpkgs/pkgs/tools/misc/qarma/default.nix:1:1 called without required argument 'qtx11extras'

and

> $ nix-build ./nixpkgs/ -A qarma
> error: anonymous function at /tmp/milahu/nixpkgs-qt6-init/nixpkgs/pkgs/tools/misc/qarma/default.nix:1:1 called without required argument 'qt5compat'

SuperSandro2000 commented on pkgs/tools/misc/qarma-for-qt6/default.nix:
```diff
@@ -0,0 +1,69 @@
+
+mkDerivation rec {
+  pname = "qarma";
+  version = "2021-10-05";
```
```suggestion
  version = "unstable-2021-10-05";
```

milahu replied:
its not "unstable" ... the maintainer just does not make new releases https://github.com/luebking/qarma/issues/31

SuperSandro2000 commented on pkgs/development/libraries/qt-6/modules/qtquick3d.nix:
```diff
@@ -0,0 +1,12 @@
+qtModule {
+  pname = "qtquick3d";
+  qtInputs = [ qtbase qtdeclarative ];
+  buildInputs = [ openssl openssl.dev ];
```
openssl.out or openssl.bin? You are sure this works for cross compile? This looks weird.

SuperSandro2000 commented on pkgs/development/libraries/qt-6/modules/qtwebengine.nix:
```diff
@@ -0,0 +1,375 @@
+, lib
+, stdenv
+, fetchpatch
+, version ? null
```
```suggestion
```
unused

SuperSandro2000 commented on pkgs/development/libraries/qt-6/modules/qtwebengine.nix:
```diff
@@ -0,0 +1,375 @@
+
+  # disable ninja line-clearing
+  preBuild = ''
+    export TERM=dumb
```
Can we move this to ninja?

milahu replied:
fixed in #149810

SuperSandro2000 commented on pkgs/applications/graphics/qimgv/default.nix:
```diff
@@ -22,23 +23,31 @@ mkDerivation rec {
   ];
 
-  buildInputs = [
+  propagatedBuildInputs = [
```
Do we really need all of this?

SuperSandro2000 commented on pkgs/development/libraries/qt-6/modules/qtbase.nix:
```diff
@@ -0,0 +1,555 @@
+        ++ lib.optional withGtk3 gtk3
+    )
+    ++ lib.optional developerBuild gdb
+    ++ lib.optional (cups != null) cups
+    ++ lib.optional (libmysqlclient != null) libmysqlclient
+    ++ lib.optional (postgresql != null) postgresql
```
Overwriting things with null in inputs is a relict from before callPackage. Can we please use options for such things?

milahu committed:
qt6: cleanup: remove experimental code

milahu committed:
qtwayland: pkg-config to native inputs

milahu committed:
qt6.qtwebengine: rm version attr

milahu committed:
qt6: patch-cmake-files.sh: sh -> bash, comments

milahu commented on pkgs/development/libraries/openvg-headers/default.nix:
```diff
@@ -0,0 +1,28 @@
+    homepage    = "https://www.khronos.org/registry/OpenVG/";
+    platforms   = platforms.linux;
+    license     = licenses.asl20; # assume same license as vulkan-headers
+    #maintainers = [ maintainers.ralith ];
```
moving openvg to separate PR, not-yet needed for qt6

milahu committed:
remove: openvg openvg-headers FindOpenVG.cmake

milahu committed:
cleanup: ninja inputs

prusnak commented:
Trying to build `qt6.qtwebengine`:

```
$ nix-build --show-trace . -A qt6.qtwebengine
error: while evaluating 'callPackageWith' at /mnt/fast/tmp/work/nixpkgs/lib/customisation.nix:117:35, called from /mnt/fast/tmp/work/nixpkgs/pkgs/development/libraries/qt-6/6.2/default.nix:139:21:
while evaluating 'makeOverridable' at /mnt/fast/tmp/work/nixpkgs/lib/customisation.nix:67:24, called from /mnt/fast/tmp/work/nixpkgs/lib/customisation.nix:121:8:
anonymous function at /mnt/fast/tmp/work/nixpkgs/pkgs/development/libraries/qt-6/modules/qtwebengine.nix:1:1 called with unexpected argument 'version', at /mnt/fast/tmp/work/nixpkgs/lib/customisation.nix:69:16
```

prusnak commented:
Trying to build `qt6.qtbase` on ef57a07f3276ff212f9ead0c816c3848ddab9fea - build fails with:

```
$ nix-build --show-trace . -A qt6. qtbase
...
...
'/nix/store/vci585j9ns8wf4b64wyq72jhgmvg4m7g-qtbase-6.2.2-dev/libexec' -> '/nix/store/b9h1l7vy01h9y3im2r4jsp06s64r3wvw-qtbase-6.2.2/libexec'             
patching output paths in /nix/store/vci585j9ns8wf4b64wyq72jhgmvg4m7g-qtbase-6.2.2-dev/lib/cmake ...                                                      
xargs: /nix/store/h90r8m4vg7z4rz59x0drp0cqdf19dza6-patch-cmake-files.sh: No such file or directory                                                       
builder for '/nix/store/3h5z0k9rdiw71sivd3hc5r0aq1ykfl3p-qtbase-6.2.2.drv' failed with exit code 127                                                     
error: build of '/nix/store/3h5z0k9rdiw71sivd3hc5r0aq1ykfl3p-qtbase-6.2.2.drv', '/nix/store/vxzm6xj6jhkaqh5crpxg8zjbzf26a9ra-qttools-6.2.2.drv' failed
```

milahu committed:
qt6.qtbase: add patch-cmake-files.sh to buildInputs

milahu committed:
qt6: rm version from callpackage qtwebengine

milahu committed:
qt6: qtModule: add patch-cmake-files.sh to buildInputs

milahu commented:
> ```diff
> +, version
> ```

version is set in qtModule.nix

```nix
{
  version = args.version or srcs.${pname}.version;
```

milahu committed:
qt6: platforms: unix -> linux

milahu committed:
revert 2a4c3b4634cd9b91bb93d9a55593287b64f37eeb

milahu committed:
qt6: fix eval checks

prusnak commented:
Trying to build 6bcb44cc48e5a497a305da0eb5401ec3ebb29524:

```
$ nix-build . -A qt6.qtbase
these derivations will be built:
  /nix/store/sgyw5p7djmi7m0i4907igkz71vsmy222-qtbase-6.2.2.drv
building '/nix/store/sgyw5p7djmi7m0i4907igkz71vsmy222-qtbase-6.2.2.drv'...
/nix/store/h90r8m4vg7z4rz59x0drp0cqdf19dza6-patch-cmake-files.sh: /nix/store/h90r8m4vg7z4rz59x0drp0cqdf19dza6-patch-cmake-files.sh: /usr/bin/env: bad interpreter: No such file or directory
builder for '/nix/store/sgyw5p7djmi7m0i4907igkz71vsmy222-qtbase-6.2.2.drv' failed with exit code 126
error: build of '/nix/store/sgyw5p7djmi7m0i4907igkz71vsmy222-qtbase-6.2.2.drv' failed
```

milahu committed:
partially revert 96fed0964bbbb2e29d5169dc9caa4d3dfe2e8d2c

milahu commented on pkgs/development/libraries/qt-6/patch-cmake-files.sh:
```diff
@@ -0,0 +1,59 @@
+#! /bin/sh
```
why...? its a posix shell script

building a bash script is more complex = [substitute](https://nixos.org/manual/nixpkgs/stable/#fun-substitute)

symphorien replied:
/bin/sh is not available in the sandbox anyway...

milahu replied:
works for me ...

milahu commented on pkgs/development/python-modules/pyside6/test.nix:
```diff
@@ -0,0 +1,88 @@
+        version='${version}',
+        scripts=['${pname}'],
+
+        # FIXME
+        #install_requires=['PySide6'],
+        # ERROR: Could not find a version that satisfies the requirement PySide6
```
fixme

Artturin replied:
@milahu i fixed this in the latest push by doing the same thing arch does in pyside6 and shiboken6 https://github.com/archlinux/svntogit-packages/blob/e1fe2063534324b8d978bb3a5e0b687f6e07f8ea/trunk/PKGBUILD#L34

milahu replied:
thanks!

weird they still use *.egg format ... but ok

Artturin commented:
```
nix build ".#python3Packages.pyside6"
qtbase> /nix/store/y1mkklskqfnqdk562dggbhhvswaz78gi-patch-cmake-files.sh: no files in argv -> ignore
qtbase> /nix/store/y1mkklskqfnqdk562dggbhhvswaz78gi-patch-cmake-files.sh: line 17: 0: command not found
```

milahu commented:
works for me 0__o

<details>

```
$ nix build ./nixpkgs#qt6.qtbase --print-build-logs
[ ... ]
qtbase> patching output paths in /nix/store/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev/lib/cmake ...
qtbase> /nix/store/1zigq9zhbx4avmx2nmxn11lakfxiqf3k-patch-cmake-files.sh eval qtCompatVersion=6.2.2
qtbase> /nix/store/1zigq9zhbx4avmx2nmxn11lakfxiqf3k-patch-cmake-files.sh eval QT_MODULE_NAME=QTBASE
qtbase> /nix/store/1zigq9zhbx4avmx2nmxn11lakfxiqf3k-patch-cmake-files.sh eval NIX_OUT_PATH=/nix/store/8mla2zz46l6lfgb9ma9bb5z70h8sj26g-qtbase-6.2.2
qtbase> /nix/store/1zigq9zhbx4avmx2nmxn11lakfxiqf3k-patch-cmake-files.sh eval NIX_DEV_PATH=/nix/store/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev
qtbase> /nix/store/1zigq9zhbx4avmx2nmxn11lakfxiqf3k-patch-cmake-files.sh eval NIX_BIN_PATH=/nix/store/yzjasqlsl9sqs60p9rbrmn3mcxnh25hi-qtbase-6.2.2-bin
qtbase> /nix/store/1zigq9zhbx4avmx2nmxn11lakfxiqf3k-patch-cmake-files.sh eval REGEX_FILE=/nix/store/iyg8dphi31la9vg7v9nb7dfxgyc6lg56-patch-cmake-files.regex.>
qtbase> /nix/store/1zigq9zhbx4avmx2nmxn11lakfxiqf3k-patch-cmake-files.sh: patching 536 files
qtbase> /nix/store/1zigq9zhbx4avmx2nmxn11lakfxiqf3k-patch-cmake-files.sh: debug:
qtbase>   REGEX_FILE = /nix/store/iyg8dphi31la9vg7v9nb7dfxgyc6lg56-patch-cmake-files.regex.diff
qtbase>   qtCompatVersion = 6.2.2
qtbase>   QT_MODULE_NAME = QTBASE
qtbase>   NIX_OUT_PATH = /nix/store/8mla2zz46l6lfgb9ma9bb5z70h8sj26g-qtbase-6.2.2
qtbase>   NIX_DEV_PATH = /nix/store/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev
qtbase>   NIX_BIN_PATH = /nix/store/yzjasqlsl9sqs60p9rbrmn3mcxnh25hi-qtbase-6.2.2-bin
qtbase> /nix/store/1zigq9zhbx4avmx2nmxn11lakfxiqf3k-patch-cmake-files.sh: build regex
qtbase> /nix/store/1zigq9zhbx4avmx2nmxn11lakfxiqf3k-patch-cmake-files.sh: regex:
qtbase> s/^# Compute the installation prefix relative to this file\.\n.*?set\(_IMPORT_PREFIX ""\)\nendif\(\)/set(_QTBASE_NIX_OUT "\/nix\/store\/8mla2zz46l6lfgb9ma9bb5z70h8sj26g-qtbase-6.2.2")\nset(_QTBASE_NIX_DEV "\/nix\/store\/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev")\nset(_QTBASE_NIX_BIN "\/nix\/store\/yzjasqlsl9sqs60p9rbrmn3mcxnh25hi-qtbase-6.2.2-bin")/s;
qtbase> s/\${_IMPORT_PREFIX}\/(\.\/)?include/\${_QTBASE_NIX_DEV}\/include/g;
qtbase> s/\${_IMPORT_PREFIX}\/(\.\/)?libexec/\${_QTBASE_NIX_OUT}\/libexec/g;
qtbase> s/\${_IMPORT_PREFIX}\/(\.\/)?lib/\${_QTBASE_NIX_OUT}\/lib/g;
qtbase> s/\${_IMPORT_PREFIX}\/(\.\/)?plugins/\${_QTBASE_NIX_BIN}\/lib\/qt-6.2.2\/plugins/g;
qtbase> s/\${_IMPORT_PREFIX}\/(\.\/)?bin/\${_QTBASE_NIX_DEV}\/bin/g;
qtbase> s/\${_IMPORT_PREFIX}\/(\.\/)?mkspecs/\${_QTBASE_NIX_DEV}\/mkspecs/g;
qtbase> s/\${_IMPORT_PREFIX}\/(\.\/)?qml/\${_QTBASE_NIX_OUT}\/lib\/qt-6.2.2\/lib\/qml/g;
qtbase> s/set\(_IMPORT_PREFIX\)/set(_QTBASE_NIX_OUT)\nset(_QTBASE_NIX_DEV)\nset(_QTBASE_NIX_BIN)/g;
qtbase> s/\${QtBase_SOURCE_DIR}\/libexec/\${QtBase_BINARY_DIR}\/libexec/g;
qtbase> s/\${QT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX}\/\${INSTALL_LIBEXECDIR}/\/nix\/store\/8mla2zz46l6lfgb9ma9bb5z70h8sj26g-qtbase-6.2.2\/libexec/g;
qtbase> s/\${QT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX}\/\${INSTALL_BINDIR}/\/nix\/store\/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev\/bin/g;
qtbase> s/\${QT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX}\/\${INSTALL_DOCDIR}/\/nix\/store\/8mla2zz46l6lfgb9ma9bb5z70h8sj26g-qtbase-6.2.2\/share\/doc/g;
qtbase> s/\${QT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX}\/\${INSTALL_LIBDIR}/\/nix\/store\/8mla2zz46l6lfgb9ma9bb5z70h8sj26g-qtbase-6.2.2\/lib/g;
qtbase> s/\${QT_BUILD_INTERNALS_RELOCATABLE_INSTALL_PREFIX}\/\${INSTALL_MKSPECSDIR}/\/nix\/store\/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev\/mkspecs/g;
qtbase> s/\${CMAKE_CURRENT_LIST_DIR}\/\.\.\/mkspecs/\/nix\/store\/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev\/mkspecs/g;
qtbase> s/\${CMAKE_CURRENT_BINARY_DIR}\/mkspecs/\/nix\/store\/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev\/mkspecs/g;
qtbase> s/(  INTERFACE_INCLUDE_DIRECTORIES )/  INTERFACE_LINK_DIRECTORIES "\${_QTBASE_NIX_OUT}\/lib"\n\1/gs;
qtbase> s/\${CMAKE_CURRENT_LIST_DIR}\/\.\.\/mkspecs/\/nix\/store\/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev\/mkspecs/g;
qtbase> s/\${CMAKE_CURRENT_BINARY_DIR}\/mkspecs/\/nix\/store\/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev\/mkspecs/g;
qtbase> patching output paths in /nix/store/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev/lib/cmake done
qtbase> writing /nix/store/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev/bin/qt.conf
qtbase> verify that all _IMPORT_PREFIX are replaced in /nix/store/41spflm6n8lfgky7k37c71fl2vv3bi6y-qtbase-6.2.2-dev/lib/cmake
qtbase> verify that all _IMPORT_PREFIX are replaced done

$ nix-build nixpkgs/ -A qt6.qtbase 
/nix/store/8mla2zz46l6lfgb9ma9bb5z70h8sj26g-qtbase-6.2.2

$ ls /nix/store/8mla2zz46l6lfgb9ma9bb5z70h8sj26g-qtbase-6.2.2/lib/
libQt6Concurrent.prl                   libQt6Network.so.6.2.2
libQt6Concurrent.so                    libQt6OpenGL.prl
libQt6Concurrent.so.6                  libQt6OpenGL.so
libQt6Concurrent.so.6.2.2              libQt6OpenGL.so.6
libQt6Core.prl                         libQt6OpenGL.so.6.2.2
libQt6Core.so                          libQt6OpenGLWidgets.prl
libQt6Core.so.6                        libQt6OpenGLWidgets.so
libQt6Core.so.6.2.2                    libQt6OpenGLWidgets.so.6
libQt6DBus.prl                         libQt6OpenGLWidgets.so.6.2.2
libQt6DBus.so                          libQt6PrintSupport.prl
libQt6DBus.so.6                        libQt6PrintSupport.so
libQt6DBus.so.6.2.2                    libQt6PrintSupport.so.6
libQt6DeviceDiscoverySupport.a         libQt6PrintSupport.so.6.2.2
libQt6DeviceDiscoverySupport.prl       libQt6Sql.prl
libQt6EglFSDeviceIntegration.prl       libQt6Sql.so
libQt6EglFSDeviceIntegration.so        libQt6Sql.so.6
libQt6EglFSDeviceIntegration.so.6      libQt6Sql.so.6.2.2
libQt6EglFSDeviceIntegration.so.6.2.2  libQt6Test.prl
libQt6EglFsKmsSupport.prl              libQt6Test.so
libQt6EglFsKmsSupport.so               libQt6Test.so.6
libQt6EglFsKmsSupport.so.6             libQt6Test.so.6.2.2
libQt6EglFsKmsSupport.so.6.2.2         libQt6Widgets.prl
libQt6FbSupport.a                      libQt6Widgets.so
libQt6FbSupport.prl                    libQt6Widgets.so.6
libQt6Gui.prl                          libQt6Widgets.so.6.2.2
libQt6Gui.so                           libQt6XcbQpa.prl
libQt6Gui.so.6                         libQt6XcbQpa.so
libQt6Gui.so.6.2.2                     libQt6XcbQpa.so.6
libQt6InputSupport.a                   libQt6XcbQpa.so.6.2.2
libQt6InputSupport.prl                 libQt6Xml.prl
libQt6KmsSupport.a                     libQt6Xml.so
libQt6KmsSupport.prl                   libQt6Xml.so.6
libQt6Network.prl                      libQt6Xml.so.6.2.2
libQt6Network.so                       metatypes
libQt6Network.so.6

$ cat /etc/nix/nix.conf 
# WARNING: this file is generated from the nix.* options in
# your NixOS configuration, typically
# /etc/nixos/configuration.nix.  Do not edit it!
build-users-group = nixbld
max-jobs = 40
cores = 32
sandbox = true
extra-sandbox-paths = 
substituters = https://cache.nixos.org/
trusted-substituters = 
trusted-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
auto-optimise-store = true
require-sigs = true
trusted-users = root @wheel jon nixpkgs-update tim jtojnar hydra-queue-runner
allowed-users = *
builders =
system-features = nixos-test benchmark big-parallel kvm
sandbox-fallback = false
experimental-features = nix-command flakes
keep-outputs = true
keep-derivations = true
```

</details>

ideally "patching output paths" should run before configurePhase ...

if you want to debug this, in `nixpkgs/pkgs/development/libraries/qt-6/modules/qtbase.nix`
set `splitBuildInstallEnabled = true;`
and copy-paste the `postFixup` to the second `stdenv.mkDerivation`

milahu committed:
qt6: remove patch-cmake-files-sh from buildInputs (thx alex)

milahu committed:
qt6: update patch-cmake-files.sh

milahu committed:
qt6: add unit test for patch-cmake-files.sh

Kranzes commented:
Perhaps squash the commits so it isn't 120+ commits 😄

davidlenfesty commented:
I just ran this today and it fails to build for me.

Tried to build a nix shell with:

```nix
{ pkgs ? import (builtins.fetchGit {                                                                                                                          
    name = "nixpkgs-qt6";                                                                                                                                     
    url = "https://github.com/milahu/nixpkgs/";                                                                                                               
    ref = "refs/heads/qt-6-init";                                                                                                                             
    rev = "5788ed07598872509240100f722aad4eef88e2ce";                                                                                                         
  }) {}                                                                                                                                                       
}:                                                                                                                                                            
                                                                                                                                                              
with pkgs;                                                                                                                                                    
mkShell {                                                                                                                                                     
  buildInputs = with python310Packages; [                                                                                                                     
    pyside6                                                                                                                                                   
  ];                                                                                                                                                          
}
```

Results in:
```
these derivations will be built:
  /nix/store/65qxjv59wla2yysw8wrfdk4f69z7gdyp-pyside6-6.2.2.drv
building '/nix/store/65qxjv59wla2yysw8wrfdk4f69z7gdyp-pyside6-6.2.2.drv'...
qtPreHook
unpacking sources
unpacking source archive /nix/store/ppwj8278d7yzqcvz909crdxdjbiafz2f-pyside-setup-opensource-src-6.2.2.tar.xz
source root is pyside-setup-opensource-src-6.2.2
setting SOURCE_DATE_EPOCH to timestamp 1638409702 of file pyside-setup-opensource-src-6.2.2/sources/shiboken6/data/docgenerator.1
patching sources
applying patch /nix/store/3j0ri158pik1c9kzzf3b0ak1g9glzr0x-dont_ignore_optional_modules.patch
patching file sources/pyside6/cmake/PySideHelpers.cmake
configuring
no configure script, doing nothing
building
build flags: -j16 -l16
ninja: error: loading 'build.ninja': No such file or directory
builder for '/nix/store/65qxjv59wla2yysw8wrfdk4f69z7gdyp-pyside6-6.2.2.drv' failed with exit code 1
error: build of '/nix/store/65qxjv59wla2yysw8wrfdk4f69z7gdyp-pyside6-6.2.2.drv' failed
```

Still pretty new with nix so this may also be me doing something wrong and be completely irrelevant.

milahu commented:
please try 9cda63094a0d6b2918aebbf73148bac3543d33c4. [pyside6](https://download.qt.io/official_releases/QtForPython/pyside6/PySide6-6.2.2-src/) has only `CMakeLists.txt` but no `build.ninja`

meanwhile there is qt 6.2.3 ...

Artturin commented:
@milahu i have rebased this branch on 'master', there were some conflicts in all-packages.nix because you had accidentally commited some unrelated edits but i fixed them properly by rebase editing and amending the fixes

would you like for me to do a force push to keep this branch in a maintainable state?

milahu commented:
yes please, thanks for the rebase

`@all` please dont wait for me to fix this PR. most parts are working, thats good enough for me

Artturin committed:
qt6: xlibs -> xorg

xlibs is a throw alias

Artturin committed:
shiboken6: init at 6.2.0

Artturin committed:
python3Packages.pyside6{,-tools}: init at 6.2.0

prusnak commented:
It seems that Qt6 6.4.0 will re-introduce pkg-config files: https://bugreports.qt.io/browse/QTBUG-86080/ and these will be probably backported to Qt6 6.3.1.

peterhoeg commented:
Qt 6.3 was released y'day: https://www.qt.io/blog/qt-6.3-released

peterhoeg commented:
@ttuegel , do you have time to take a look at this?

peterhoeg commented:
And just to add this - I understand that 6.2 is an LTS release, but KDE as eventually the main downstream consumer of Qt6 is presumably going to be requiring the latest and thus 6.3 is likely to be a requirement for that.

Of course, there is nothing preventing us from carrying both the latest LTS as well as the latest and greatest.

peterhoeg commented on pkgs/applications/audio/strawberry/default.nix:
```diff
@@ -25,7 +25,8 @@
 , qtbase
-, qtx11extras
+, qtx11extras ? null # qt5
+, qt5compat ? null # qt6
```
`qt5compat` isn't used

peterhoeg commented on pkgs/applications/audio/strawberry/default.nix:
```diff
@@ -62,6 +63,8 @@ mkDerivation rec {
     taglib
     qtbase
     qtx11extras
+  ] ++ lib.optionals (lib.versions.major qtbase.version == "5") [
```
I wondering if we should introduce `isQt5` and `isQt6` helpers like what we have for Python as this check (or a variation of it) is most likely going to be littered all over the place.

peterhoeg commented on pkgs/development/libraries/qt-6/README.md:
```diff
@@ -0,0 +1,81 @@
+# Qt 5 Maintainer's Notes
```
There's a few references to Qt 5 that should Qt 6.

milahu commented:
> @milahu are there any merge blockers?

"someone" (sorry, i have no energy for this) would need to squash all the qt6 commits
and separate them from the "switch app to qt6" (etc) commits
(also see my todo list in the first post)

apart from that ... not that i know of. it's "basically working"

> I'd be glad to merge this before the feature freeze window closes if possible.

yepp. better than "death by perfectionism"

Artturin commented:
Adding qt6 isn't a breaking change so it can be merged anytime and backported

milahu commented on pkgs/development/libraries/qt-6/packaging-qt6.md:
```diff
@@ -0,0 +1,159 @@
+either dependencies are missing,
+or the feature is disabled by default.
+
+qt5 used `configure`. for example `./configure -system-ffmpeg`
+
+qt6 uses cmake, so we need ...
```
iirc, some options go to configure, other options go to cmake

ive added a script to the readme to list the available cmake options

oxalica commented:
Note: `tdesktop` 3.6.3 requires Qt 6 to build now, thus is blocked by this PR.
https://github.com/telegramdesktop/tdesktop/releases/tag/v3.6.3

K900 commented:
Anything holding this back still? tdesktop 3.7.0 also requires Qt6 now, and the current version requires some jank to keep going (see https://github.com/NixOS/nixpkgs/pull/169180).

NickCao commented:
I think the biggest blocker is whether it builds.

NickCao commented:
Just picked this up https://github.com/NickCao/nixpkgs/tree/qt6 and bumped it to qt 6.3.0, surprisingly most of the things builds with minor fixes (with qtwebengine as the only exception). tdesktop 3.7.0 builds and works fine with this.

Kranzes commented:
https://github.com/NickCao/nixpkgs/blob/deeb74157bf5142c848c8c9903072ffa0ea2362a/pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix#L162 You forgot about this?

NickCao commented:
Still having some trouble in getting IME (fcitx5) to work.

Kranzes commented:
> Just picked this up https://github.com/NickCao/nixpkgs/tree/qt6 and bumped it to qt 6.3.0, surprisingly most of the things builds with minor fixes (with qtwebengine as the only exception). tdesktop 3.7.0 builds and works fine with this.

I also tested tdesktop 3.7.0 from your branch, works fine.

milahu commented:
@NickCao thx : )

im curious how 8b2eed21b7ecdb8f788810f21ea927c0cabea8c5 works ...

iirc, my goal was to make the qt libs usable in cmake projects
by adding the line

```
  INTERFACE_LINK_DIRECTORIES "\${_((QT_MODULE_NAME))_NIX_OUT}/lib"
```

before the line

```
  INTERFACE_INCLUDE_DIRECTORIES ...
```

in the generated `*.cmake` files
otherwise cmake would say "package found but files are missing"

so, in my remote diagnosis, your version should break the tdesktop build, but apparently it works?

maybe this was fixed between qt 6.2 and 6.3
how do your generated `*.cmake` files look when you `grep -C10` for `INTERFACE_INCLUDE_DIRECTORIES`?

ttuegel commented:
> The lack of this particular patch should be the root cause: https://github.com/NickCao/nixpkgs/blob/qt6/pkgs/development/libraries/qt-5/5.12/qtbase.patch.d/0010-qtbase-qtpluginpath.patch

This patch caused a lot of headaches for Qt 5 and in the strongest way possible, I would encourage us **not** to make that mistake again. Maintainers should set up the correct wrappers for their applications.

NickCao commented:
> im curious how [8b2eed2](https://github.com/NixOS/nixpkgs/commit/8b2eed21b7ecdb8f788810f21ea927c0cabea8c5) works ...
 
Me myself don't know how either, I removed that line simply because it was patching too much, breaking unrelated cmake files specifically the one for libdrm vendoered in qtbase. And somehow, it just works.

Artturin commented:
Still needs lots of cleanup and resyncing the improvements to the qt5 packaging made during the time this pr has been open

The `? null`s should he replaced with something better too

NickCao commented:
Tinkered a little bit around the cmake patching hacks, found that it seem all the patched files are named SomeQtConponentTargets.cmake and is generated by cmake. Can we just patch the template from which the cmake files are generated and solve this once and for all?

NickCao commented:
By applying a questionable and over fitting patch atop cmake, I managed to cut the hacky list of regex by half.
https://github.com/NickCao/nixpkgs/commit/8cd585437ca1f895b8886dba20d9e677a2bb8a8a

NickCao commented:
By further neglecting support for qmake, the final approach is now clean and clear:
https://github.com/NickCao/nixpkgs/commit/f5aa39f7770cefbcdb56d2ae0acea3252147095d

NickCao commented:
Just noticed that the qt plugins depends on the main output in their rpaths, so what's the point of splitting them into another output?

NickCao commented:
Also as the coexisting multiple minor qt versions approach has never really worked out, dropping that also looks like a nice choice for me.

milahu commented:
> most of the things builds with minor fixes (with qtwebengine as the only exception)

what's the issue with qtwebengine?

NickCao commented:
> 
`./third_party/webgpu-cts/src/tools/deno: unsupported interpreter directive "#!/usr/bin/env -S deno run --unstable --allow-read --allow-write --allow-env --no-check" (set dontPatchShebangs=1 and handle shebang patching yourself)`

milahu commented:
oh .. stupid patchShebangs is confused by the `-S` option

https://github.com/NixOS/nixpkgs/blob/bf27ea9cdb0c7cdc1f7aee1a29914b83684dac03/pkgs/build-support/setup-hooks/patch-shebangs.sh#L64-L71

you could try

```nix
{
  postPatch = ''
    sed -i.bak 's|^#!/usr/bin/env -S deno|#! ${deno}/bin/deno|' ./third_party/webgpu-cts/src/tools/deno
    diff -u ./third_party/webgpu-cts/src/tools/deno{.bak,} || true; exit 1 # debug
  '';
```

milahu commented:
closing in favor of #169296 

keeping the branch [qt-6-init](https://github.com/milahu/nixpkgs/tree/qt-6-init)
